src/pull-request-core.ts

```diff
@@ -0,0 +1,234 @@
+                    octokit.git.createCommit({
+                        owner,
+                        repo,
+                        message: commitMessage,
+                        tree: treeSha,
+                        author: {
+                            name: authorName,
+                            email: authorEmail,
+                            date: new Date().toISOString()
+                        },
+                        parents: [data.sha]
+                    })
+                )
+                .then(({ data }) =>
+                    // update the branch ref to point to the created commit
+                    octokit.git.updateRef({
+                        owner,
+                        repo,
+                        ref: 'heads/' + commitBranch,
+                        sha: data.sha
+                    })
+                )
+                .then(({ data }) => data)
+        );
+    }
+
+    /**
+     * Creates a tree from the provided commitBranchSha ref for sourceFiles and returns the sha.
+     */
+    function createTree(
+        owner: string,
+        repo: string,
+        commitBranchSha: string,
+        sourceFiles: SourceFile[]
+    ): Promise<string[]> {
+        const tree = sourceFiles.map<Octokit.GitCreateTreeParamsTree>(
+            ({ path, content }) => ({
+                path,
+                content,
+                type: 'blob',
+                mode: '100644'
+            })
+        );
+        return Promise.resolve(
+            octokit.git
+                .createTree({
+                    owner,
+                    repo,
+                    tree,
+                    base_tree: commitBranchSha
+                })
+                .then(({ data }) => [data.sha, commitBranchSha])
+        );
+    }
+
+    /** Creates a a commitBranch from baseBranch and returns the sha. */
+    function createCommitBranch(
+        owner: string,
+        repo: string,
+        baseBranch: string,
+        commitBranch: string
+    ): Promise<string> {
+        return Promise.resolve(
+            octokit.git
+                .getRef({
+                    owner,
+                    repo,
+                    ref: 'heads/' + baseBranch
+                })
+                .then(({ data }) =>
+                    octokit.git.createRef({
+                        owner,
+                        repo,
+                        ref: 'refs/heads/' + commitBranch,
+                        sha: data.object.sha
+                    })
+                )
+                .then(({ data }) => data.object.sha)
+        );
+    }
+
+    /**
+     * Returns the commit branch sha ref (the one we will PR) if it exists, or creates it from the base branch and returns it.
+     */
+    function getCommitBranchSha(
+        owner: string,
+        repo: string,
+        baseBranch: string,
+        commitBranch: string
+    ): Promise<string> {
+        return new Promise((resolve, reject) => {
```

Comment:
You almost never need the Promise constructor, unless you need to convert some other callback API to Promises. When you do need to use it, you should minimize the code inside it. Here you already have a Promise at hand so you can just return and chain that (or use async/await).

If you were looking for a way to reject the Promise, you just need to `throw` inside a handler or async function.

Reply:
Makes sense. Is `throw new Error(...)` a reasonable thing or are there better conventions for error handling?

Reply:
`throw new Error()` is the way to go for errors.